### PR TITLE
Open commits performance changes

### PR DESCRIPTION
**Pull request by brycemcanally**
This PR includes a set of changes to improve performance when dealing with open commit chains. 

**The user facing changes are:**
- Open commits with an unfinished base commit (most recent ancestor commit that isn't errored) are no longer readable. This only impacts output repos with parallel jobs since we already prevent the creation of a commit without a finished parent in source repos.
- Errored commits (commits associated with a killed / failed job) are no longer compacted. They are still readable, but the performance will be worse since they are not compacted. This seems like the right tradeoff in most cases and it simplifies the above change since a job can be killed before its base commit is finished.
- Repo sizes are based on the most recent commit in master that is finished (rather than open).
- ~~Commits that are not readable due to this change will have a sentinel value set for the `SizeBytesUpperBound` field (TODO)~~ (We are going to leave the size as 0 until it is in a readable state).

The problem we are hitting in a few different ways is that we are creating n file sets for each commit operation where n is the number of open ancestor commits. With list operations this can end up being n^2, which creates a lot of garbage in the system that further degrades performance. Another issue that is somewhat related is that commit finishing order is based on an updated at timestamp, which can be the same for multiple commits when a pipeline is stopped. This scenario causes the commits to be finished in a pathological order which results with the first issue and more seriously a lot of duplicate compaction work. The primary change to address these issues is a new invariant that prevents us from getting a file set for a commit with an unfinished base commit. Having this invariant ensures that there is no way to hit these problems anywhere in the system. The changes in this PR are the new invariant and the corresponding changes required to satisfy the invariant.

### CODE REVIEW

File: src/server/pfs/pfs.go [hunk 1]
```diff
@@ -253,6 +263,7 @@ var (
 	fileNotFoundRe            = regexp.MustCompile(`file .+ not found`)
 	outputCommitNotFinishedRe = regexp.MustCompile("output commit .+ not finished")
 	commitNotFinishedRe       = regexp.MustCompile("commit .+ not finished")
```
Comment (PFedak) [inline src/server/pfs/pfs.go:265]: This other one isn't actually used, which is good as it matches both of the others. It'd be nice if there's a way to get rid of these regex entirely after that grpc status change

File: src/server/pfs/server/master.go [hunk 2]
```diff
@@ -122,7 +123,7 @@ func (d *driver) finishRepoCommits(ctx context.Context, compactor *compactor, re
 			return nil
 		}
 		commit := commitInfo.Commit
-		return miscutil.LogStep(fmt.Sprintf("finishing commit %v", commit.ID), func() error {
+		return miscutil.LogStep(fmt.Sprintf("finishing commit %v", commit), func() error {
```
Comment (PFedak) [inline src/server/pfs/server/master.go:126]: thanks, I keep meaning to fix this

File: src/server/pfs/server/driver.go [hunk 3]
```diff
@@ -1173,8 +1176,9 @@ func (d *driver) listCommit(
 					ci = cis[len(cis)-1-i]
 				}
 				var err error
+				// TODO: Set SizeBytesUpperBound to a sentinel value when the base is not finished.
 				ci.SizeBytesUpperBound, err = d.commitSizeUpperBound(ctx, ci.Commit)
-				if err != nil {
+				if err != nil && !pfsserver.IsBaseCommitNotFinishedErr(err) {
```
Comment (PFedak) [inline src/server/pfs/server/driver.go:1181]: could use `errors.Is` as it's not crossing a gRPC boundary

**Review by PFedak — APPROVED** [review]
LGTM, glad this was mostly localized. I think we have other tests that cover cases with multiple failed commits so we probably don't need an explicit one

File: src/server/pfs/server/driver.go (hunk 3 above)
Comment (brycemcanally) [reply src/server/pfs/server/driver.go:1181]: Yeah, I just went for consistency here. Also, something that I didn't know until recently is that `errors.Is` is a type and value check and `errors.As` is a type check. So, we should use `errors.As` when we only care about the type.

File: src/server/pfs/pfs.go (hunk 1 above)
Comment (brycemcanally) [reply src/server/pfs/pfs.go:265]: Yeah, I'm going to go ahead and make the base commit not finished regex not matchable by the not finished regex, but we should eventually come up with a better way to handle the types than regex matching.

**Pull request merged**